### PR TITLE
Django REST APIs for Overview page

### PR DESCRIPTION
**Pull request by hooshooo**
### Why:
Django REST APIs for Overview page

### What:
1. add some basic RESTful APIs for the front-end.
2. add soem customized APIs.

### Checklist
- [x] Finished My Changes
- [ ] Automated Unit Tests

### CODE REVIEW

File: script/app/settings/dev.py [hunk 1]
```diff
@@ -1,3 +1,9 @@
 from .base import *
 
 # override the settings from base.py
+DATABASES = {
+    'default': {
+        'ENGINE': 'django.db.backends.sqlite3',
```
Comment (jongoncalves) [inline script/app/settings/dev.py]: Let's setup postgres instead of sqlite. Will make managing this in AWS via RDS a lot simpler.

File: script/app/urls.py [hunk 2]
```diff
@@ -1,23 +1,6 @@
...
-    1. Import the include() function: from django.urls import include, path
-    2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
-"""
 from django.contrib import admin
 from django.urls import path, include
-from app import views
 
 urlpatterns = [
-    path('', views.index, name='index'),
-    path('admin/', admin.site.urls),
```
Comment (jongoncalves) [inline script/app/urls.py:22]: Why are you removing the administration panel functionality (routing)?

File: script/manage.py [hunk 3]
```diff
@@ -5,7 +5,7 @@
 
 
 def main():
-    os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'app.settings')
+    os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'app.settings.dev')
```
Comment (jongoncalves) [inline script/manage.py]: Let's keep this pointed to base instead. We should all individually setup this setting locally. 
Plus we can always override any management command from `manage.py` with `--settings=app.settings.dev`

File: script/script/models.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+class ChargingStation(models.Model):
+    """Charging station by county and year-month"""
+    county = models.ForeignKey(County, on_delete=models.CASCADE)
+    year = models.IntegerField()
+    month = models.IntegerField()
+    number = models.IntegerField()
+
+    class Meta:
+        unique_together = (('county', 'year', 'month'),)
```
Comment (jongoncalves) [inline script/script/models.py]: Let's add an explicit table name to all of these. Makes thing easier to grasp if ever need to go directly to the DB.
i.e:
```
class Meta:
        db_table = 'a_random_table_name'
```

File: script/script/models.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
+from django.db import models
+
+# Create your models here.
+
+class County(models.Model):
+    """County Info"""
+    name = models.CharField(max_length=50, primary_key=True)
+    residents = models.IntegerField()
```
Comment (jongoncalves) [inline script/script/models.py]: Is this meant to be the count of residents in the county?
Wondering if there's a better name for this...when I read `residents` my mind assumes a related resident model

**Review by gcezar — APPROVED** [review]

File: script/script/models.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    # [TODO] add other basic statistics of the county
+
+
+class Energy(models.Model):
+    """Energy consumed by county and year-month"""
+    county = models.ForeignKey(County, on_delete=models.CASCADE)
+    year = models.IntegerField()
+    month = models.IntegerField()
```
Comment (jongoncalves) [inline script/script/models.py]: What's the benefit of using Integers for year and month instead of a date object?

I'm fine with int as long as there's an accompanying validator. The way it is right now would allow for corrupted values (values that would make no sense in the context of year and month)

File: script/script/models.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    class Meta:
+        unique_together = (('county', 'year', 'month'),)
+
+
+class ChargingStation(models.Model):
+    """Charging station by county and year-month"""
+    county = models.ForeignKey(County, on_delete=models.CASCADE)
+    year = models.IntegerField()
+    month = models.IntegerField()
+    number = models.IntegerField()
```
Comment (jongoncalves) [inline script/script/models.py]: What is `number`? Let's find a more descriptive name for what this field is trying to capture

File: script/script/views.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+    def delete(self, request, *args, **kwargs):
+        county_name = request.query_params.get('county', None)
+        county = County.objects.get(name=county_name)
+        year = request.query_params.get('year', None)
+        month = request.query_params.get('month', None)
+        res = Energy.objects.get(county=county, year=year, month=month)
+        res.delete()
+        serializer = self.serializer_class(res)
+        return Response(serializer.data)
```
Comment (jongoncalves) [inline script/script/views.py]: A couple of points here:
- DRF should do, out of the box, all of what you are trying to capture here. All you need to do is make the HTTP request with the appropriate verb, e.g.: `POST | PUT | DELETE | GET`
- Any overrides, if needed, should live in the serializer. There could be a few exceptions....
- Lastly, if you are putting these type of operations in the view, don't name them off of the HTTP verbs. Especially without limiting the type of request to make. There's a semantic mismatch, for example, if I make a `GET` request against the `delete` endpoint.

**Review by jongoncalves — COMMENTED** [review]
Also, what is the reasoning behind renaming `app` -> `script`?
I'm not a fan of paths that duplicate the project name, like `script/script/...` instead of `script/app/...`. The latter allowing for more definition around the components that the project script is comprised of, like and app, an algorithm, an async queue, an utility function, etc...

**Comment by hooshooo** [comment]
> async queue, an utility functio
Sorry, I didn't see any difference here. Could you please explain further about it? I think I just moved all application-related files to a new directory `script/script` and left setting and wsgi file in `script/app/`. And in the furture, we will make `script/frontend/` for our react code.

File: script/script/views.py (hunk 8 above)
Comment (hooshooo) [reply script/script/views.py]: Thanks for your comments. This is my first time using DRF, and I got very confused about the design. I wrote two versions of the apis:
1. such as `/api/county/<county_name>/year/<year>`
2. such as `/api/county?county=<county_name>&year=<year>`
I actually prefer the former although the uri could be really long if there are many params in the composite unique keys. Then, the complete apis could be like:
```
GET: /api/county/<county_name>/year/<year>
PUT: /api/county/update
POST: /api/county/create
DELETE: /api/county/delete
``` 
What do you think of these apis?

Comment (hooshooo) [reply script/script/views.py]: The GET endpoint is a little bit weird, so how about `/api/county/retrieve/name/<county_name>/year/<year>`?

File: script/script/models.py (hunk 5 above)
Comment (hooshooo) [reply script/script/models.py]: I planned to add all attributes related to county to this entity. However, I was not very clear what data we have, so I just added a `residents` field.

Comment (gcezar) [reply script/script/models.py]: for county we would need `driverID, sessionID, energy, power, arrival time, departure time, zipcode, category, subcategory`

Comment (jongoncalves) [reply script/script/models.py]: >so I just added a residents field.

Got it. And it's just a count of residents within the county I assume...

File: script/script/views.py (hunk 8 above)
Comment (jongoncalves) [reply script/script/views.py]: Thanks for the responses.

So for the GET request I think this would be a better approach:
```
GET: api/v1/county/{id}/?year=1985&year=1986&year.....

# This allows our backend to something like:

years = self.request.GET.getlist("year") 
```
[Reference](https://docs.djangoproject.com/en/2.2/ref/request-response/#django.http.QueryDict.getlist)

As far as the PUT, POST and DELETE, they don't need to be named anything, the VERB on the request will tell the api (DRF) the operation to do. All you need is supply the appropriate body and id. E.g:
```
PUT: api/v1/county/{id}/
# supply the appropriate json body to perform the update
{...}

POST: api/v1/county/
# supply the appropriate json body to perform the create
{...}

DELETE: api/v1/county/{id}/
# doesn't require anything else
```

So basically, by default, DRF's router creates detailed urls at /resource/{pk} and you GET, PUT, POST and DELETE them to retrieve, update, create and delete respectively.

File: script/app/urls.py (hunk 2 above)
Comment (hooshooo) [reply script/app/urls.py:22]: Becuase according to my understanding, we will never use admin page.

File: script/app/settings/dev.py (hunk 1 above)
Comment (hooshooo) [reply script/app/settings/dev.py]: done

File: script/manage.py (hunk 3 above)
Comment (hooshooo) [reply script/manage.py]: done

File: script/script/models.py (hunk 5 above)
Comment (hooshooo) [reply script/script/models.py]: My fault, I should have written the assumptions in the code or PR.
@gcezar Got it. Thanks! But I just want to store some statistics here.

File: script/script/views.py (hunk 8 above)
Comment (hooshooo) [reply script/script/views.py]: Thanks for your advice.
For the idea of filtering by list in GET, it is really helpful but I am going to do that in a new PR.

**Comment by hooshooo** [comment]
@jongoncalves There are some models I didn't write serialzers for becuase I want to ask your for opinions on current serializers first. If they are fine, I will add in a new PR. Thanks!

File: webserver/script/models/enums.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    Honda = 'Honda'
+    UK = 'Unknown'
+    # [TODO] add more vehicle makes
+
+
+class EVType(EnumWithChoices):
+    """EV type"""
+    PLUGIN = 'PLUGIN'
+    HYBRID = 'HYBRID'
+    UK = 'UNKNOWN'
```
Comment (jongoncalves) [inline webserver/script/models/enums.py]: I think it's worth standardizing the enums to all be uppercase and the all UK values -> UNKNOWN

File: webserver/script/serializers.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+    def validate_year(self, year):
+        if year < 2000 or year > datetime.datetime.now().year:
+            raise serializers.ValidationError("Year must be 2000~now.")
+        return year
+    
+    def validate_month(self, month):
+        if month < 1 or month > 12:
+            raise serializers.ValidationError("Month must be 1~12.")
+        return month
```
Comment (jongoncalves) [inline webserver/script/serializers.py]: These validators are better suited to be on the actual model instead of the serializer. As the project evolves, there may be different routes a client may take to update the model. Things like public api, admin panels, etc... The validation being on the model DRYes all that up.
Here's an example: https://docs.djangoproject.com/en/2.2/ref/validators/

Comment (hooshooo) [reply webserver/script/serializers.py]: Done. Thank you for your suggestions!

File: webserver/script/models/data.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+    class Meta:
+        db_table = 'script_driver'
+
+
+class Vehicle(models.Model):
+    """Vehicle"""
+    make = models.CharField(max_length=20, choices=VehicleMake.choices(), default=VehicleMake.UNKNOWN)
+    model = models.CharField(max_length=30)
+    year = models.IntegerField()
```
Comment (jongoncalves) [inline webserver/script/models/data.py]: Add the year validation here.

File: webserver/app/settings/dev.py [hunk 12]
```diff
@@ -0,0 +1,9 @@
+from .base import *
+
+# override the settings from base.py
+DATABASES = {
+    'default': {
+        'ENGINE': 'django.db.backends.sqlite3',
```
Comment (jongoncalves) [inline webserver/app/settings/dev.py]: Is there a reason this is still sqlite3?

**Comment by hooshooo** [comment]
Now, `settings/base.py` is for development; `settings/test.py` is for testing; `settings/deploy.py` is for users to use with RDS.

File: webserver/script/validators.py [hunk 13]
```diff
@@ -17,3 +17,8 @@ def validate_month(month):
 def validate_positive(value):
     if value < 0:
         raise ValidationError("Value must be a positive.")
+
+
+def validate_zipcode(zipcode):
+    if len(zipcode) != 5 or zipcode.isdigit():
```
Comment (jongoncalves) [inline webserver/script/validators.py]: Seems that since this validation is on a django `CharField` that there's no use to check if it's a digit.

Everything is looking pretty good so far 👍

File: .circleci/config.yml [hunk 14]
```diff
@@ -3,7 +3,7 @@ jobs:
   test:
     docker:
       - image: continuumio/miniconda3
-      - image: postgres:11.5
+      - image: postgres:9
```
Comment (jongoncalves) [inline .circleci/config.yml:6]: What's the reason for downgrading postgres?

File: webserver/script/validators.py (hunk 13 above)
Comment (hooshooo) [reply webserver/script/validators.py]: I just found I missed a `not` here. I just want to avoid typo from users' inputs, such as `9408g`.
And the reason why I chose to use `CharField` rather than `IntegerField` is there could be some zipcodes beginning with `0`. (I'm not sure if US has such kind of zipcodes...)

File: .circleci/config.yml (hunk 14 above)
Comment (hooshooo) [reply .circleci/config.yml:6]: It's because I'm using postgres9 for my own development environment and I want to make the environments consistent.
I don't see much difference between these two versions, but if you prefer a latest version, postgres11 should also be fine.
